6 xctr plot

lindadelacombaz opened the pull request.

maurolepore commented on R/plot_sankey.R:
```diff
@@ -0,0 +1,99 @@
+  # FIXME: This is always TRUE because the first code chunks adds the column
+  # `middle_node2`. That is, the `else` branch never runs.
+  # TODO: Replace with hasName()
+  if ("middle_node2" %in% names(data_links)) {
```
FIXME: This is always `TRUE` because the first code chunk adds the column `middle_node2`. That is, the `else` branch never runs.

maurolepore replied:
You may use `hasName()`

maurolepore commented on R/plot_sankey.R:
```diff
@@ -0,0 +1,99 @@
+  # `middle_node2`. That is, the `else` branch never runs.
+  # TODO: Replace with hasName()
+  if ("middle_node2" %in% names(data_links)) {
+    links <- data_links |>
+      select(
+        "bank",
+        source = "bank",
+        target = "middle_node2",
+        value = "amount",
+        group = "PCTR_risk_category"
+      )
```
FIXME: Is this dead code? The object `links` is defined below again so this one isn't used.

maurolepore commented on R/plot_sankey.R:
```diff
@@ -0,0 +1,99 @@
+  #   name = c("high", "low", "x", "y"),
+  #   group = ifelse(name %in% c("high", "medium", "low"), name, "other")
+  # )
+  nodes <- data.frame(
+    name = unique(c(as.character(links$source), as.character(links$target)))
+  ) |>
+    mutate(
+      group = case_when(
+        name %in% c("high", "medium", "low") ~ name,
+        TRUE ~ "other"
+      )
+    )
```
Maybe simplify with `ifelse()` and `tibble()`? e.g.:

``` r
tibble::tibble(
  name = c("high", "low", "x", "y"),
  group = ifelse(name %in% c("high", "medium", "low"), name, "other")
)
#> # A tibble: 4 × 2
#>   name  group
#>   <chr> <chr>
#> 1 high  high 
#> 2 low   low  
#> 3 x     other
#> 4 y     other
```

<sup>Created on 2023-06-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

maurolepore commented on R/plot_sankey.R:
```diff
@@ -0,0 +1,99 @@
+    Value = "value",
+    NodeID = "name",
+    # FIXME: `colourScale` breaks the plot
+    # colourScale = my_color,
```
FIXME: `colourScale` breaks the plot. It might be hard to get it right. Maybe use the default colours for now?

maurolepore commented on R/plot_sankey.R:
```diff
@@ -0,0 +1,99 @@
+  )
+  # FIXME Return invisible(data). See the tidyverse design guide and search for
+  # "side effect"
+  return(p)
```
See https://design.tidyverse.org/out-invisible.html

Here we should return `invisible(data)`. However, that makes means we are no longer able to test the plot-object. One solution is to have  an internal `plot_sankey_impl()` (implementation) that returns the plot object, and one user-facing `plot_sankey()` that calls `plot_sankey_impl()` then returns `invisible(data)`.

maurolepore commented on data-raw/toy_data.R:
```diff
@@ -0,0 +1,9 @@
+# styler: off
+toy_data <- tibble::tribble(
+  ~bank, ~amount, ~amount_of_disctinct_products, ~PCTR_risk_category,  ~tilt_sec,
```
FIXME: Can we stick to the tidyverse style guide and use `snake_case`? Maybe check with Anne (the offender is the uppercase `PCTR_*` prefix).

maurolepore commented on tests/testthat/test-plot_sankey.R:
```diff
@@ -0,0 +1,12 @@
+test_that("returns an object of the expected class", {
+  p <- plot_sankey(toy_data)
+  expect_s3_class(p, "sankeyNetwork")
+  expect_s3_class(p, "htmlwidget")
+})
```
You may test expected properties of the resulting plot like this:

``` r
devtools::load_all()
#> ℹ Loading tiltPlot

p <- plot_sankey(toy_data)
p$x$nodes$name
#> [1] "tilt_sec" "A"        "B"        "low"      "high"
p$x$nodes$group
#> [1] "other" "other" "other" "low"   "high"
```

<sup>Created on 2023-06-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

maurolepore replied:
You may test expected properties of the resulting plot using something like this:

``` r
devtools::load_all()
#> ℹ Loading tiltPlot

p <- plot_sankey(toy_data)
p$x$nodes$name
#> [1] "tilt_sec" "A"        "B"        "low"      "high"
p$x$nodes$group
#> [1] "other" "other" "other" "low"   "high"
```

<sup>Created on 2023-06-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

maurolepore reviewed (approved):
LGTM, thanks @lindadelacombaz

I pushed a few commits and made a number of comments. You may address them here or follow up in separate issues/PRs.

maurolepore commented on R/plot_sankey.R:
```diff
@@ -0,0 +1,85 @@
+      value = "amount",
+      group = "PCTR_risk_category"
+    )
+  if ("middle_node2" %in% names(data_links)) {
```
FIXME: I think the `if` branch always run (and the `else` branch never runs) because the code above adds the column `middle_node2` to `data_links`.

lindadelacombaz commented on R/plot_sankey.R:
```diff
@@ -0,0 +1,85 @@
+      source = "bank",
+      target = "PCTR_risk_category",
+      value = "amount",
+      middle_node2 = "tilt_sec"
```
@maurolepore, if there are " " added, the new columns takes the string as values added in the column. For example, middlenode2 takes "tilt_sec" as values instead of "sector_a" and "sector_b "(which is what I want) - I thought that [dplyr did not need " " anymore in the mutate() function ?](https://dplyr.tidyverse.org/reference/mutate.html)

maurolepore replied:
I'll try to find the relavant article and share it later. But note the behavour is  different for `mutate()` and `filter()` compared to `select()`, and `rename()`. 

For mutate you can use a bare name like `my_col` or `.data$my_col` while for `select()` you now use "my_col".

lindadelacombaz replied:
Okay nice ! In those lines of code it is a mutate() so I will change it into the bare name. Thanks @maurolepore.

lindadelacombaz commented:
Dear @maurolepore, 

I have addressed the comments you made thanks to your review.

I have one remaining issue that I struggle to fix, namely that if I return invisible(data), the plot is not shown, even if I call plot_sankey_impl() just before. I have thus commented the line for now and will ask for your review from the changes I made.

Thank you !!

maurolepore commented on R/plot_sankey.R:
```diff
@@ -0,0 +1,54 @@
+  plot_sankey_impl(links, nodes, my_color)
+
+  # FIXME: If I return invisible(data), then the plot does not show.
+  # return(invisible(data))
```
Yo can force printing the plot with 'print(p)' then immediately finish the function with 'invisible(data)'.

This assumes you assign your plot to 'p'. 

Note the tidyverse style guide reserves explicit 'return()' to early returns only.

maurolepore commented:
Looks great @lindadelacombaz

maurolepore commented:
@lindadelacombaz, I changed my mind about returning 'data' -- instead return 'p'.

I explored other packages for visualization and they return the plot. I think it makes sense since the primary goal of those functions is to plot. And that would allow the user to modify the plot afterwards -- as we do with ggplot2.

I'm sorry for misleading you.
